File: leetcode/0372-super-pow/Solution.java
class Solution {
    private static final int MOD = 1337;

    public int superPow(int a, int[] b) {
        long result = 1;
        for (int digit : b) {
            result = (power(result, 10) * power(a, digit)) % MOD;
        }
        return (int) result;
    }

    private long power(long base, long exp) {
        base %= MOD;
        if (base < 0) base += MOD;
        long result = 1;
        while (exp > 0) {
            if ((exp & 1) == 1) {
                result = (result * base) % MOD;
            }
            base = (base * base) % MOD;
            exp >>= 1;
        }
        return result;
    }
}

File: leetcode/0372-super-pow/solution.py
from typing import List

MOD = 1337


class Solution:
    def superPow(self, a: int, b: List[int]) -> int:
        def power(base: int, exp: int) -> int:
            base %= MOD
            result = 1
            while exp > 0:
                if exp & 1:
                    result = (result * base) % MOD
                base = (base * base) % MOD
                exp >>= 1
            return result

        result = 1
        for digit in b:
            result = (power(result, 10) * power(a, digit)) % MOD
        return result
